[FIX] 소켓 동기화 문제 해결 (#38)

dyeon-dev opened the pull request:
## 🔍 PR 요약
fix/38-merge 브랜치에 머지된 PR들을 바탕으로 소켓 동기화 버그 및 배포 오류를 해결하였습니다.
- 코드 동기화 해결 (12/19)
- 참가자 동기화 해결 (12/22)
- 배포환경에서 `ws://localhost:3000` -> `ws://223.130.160.245/` 로 요청 url 변경

- PR #40
- PR #42 
- PR #43 
- PR #33 

## 🧾 관련 이슈

> 이 PR이 관련된 이슈 번호를 명시해주세요.

- close #38 


## 🛠️ 주요 변경 사항
#### 코드 동기화 해결 (12/19)
- components/Battle/CodeEditor.tsx
  - codeEditor에서 방 입장 시 전체 상태 동기화가 되면 useRoomStore에 저장
- components/Spectator/CodeSpectator.tsx
  - CODE_CHANGE 이벤트 리스너 전달이 안되는 문제 해결
```
// 스토어에 이미 만들어진 소켓만 읽어오고, 없으면 null이라 바로 리스너 등록/emit이 실패
const client = useBattleSocketStore((state) => state.socket))

--- 아래 코드로 변경---

// socket과 connect 둘 다 받아온 뒤, 
const socket = useBattleSocketStore((state) => state.socket);
const connect = useBattleSocketStore((state) => state.connect);

// useEffect에서 socket ?? connect()로 소켓이 없으면 생성·연결까지 해 주기 때문에 항상 유효한 소켓 인스턴스(client)를 확보해 이벤트를 등록
useEffect(() => {
const client = socket ?? connect();
```

#### 참가자 동기화 해결 (12/22)
- packages/constants/socket-event.ts
  - 메인페이지쪽에서 ROOM_STATE_SYNC 이벤트로 player/spectator를 가져오던 로직을 ROOM_STATE_ROLE 이벤트로 네이밍 변경
  - -> 이벤트가 사용되는 의도가 '역할을 가져오는 이벤트'인데 '방 상태 동기화' 의미였던 ROOM_STATE_SYNC 의도와 맞지 않기 때문에 의도를 명확하게 하기 위해 변경
  - ROOM_PLAYERS 이벤트를 추가하여 방 전체 참가자 목록을 가져오기
- apps/api/src/room/room.gateway.ts
  - 서버: 방 참여자 목록 브로드캐스트 추가
- apps/web/src/stores/battleSocketStore.ts
  - 클라이언트: joinRoom 단계에서 ROOM_PLAYERS를 스토어에 반영하도록 추가

#### 배포환경에서 요청 url 변경
- apps/api/Dockerfile
  - 배포환경 빌드 시 산출물 경로 안맞는 문제 해결
  - `CMD ["node", "dist/main.js"]` -> CMD ["node", "dist/apps/api/src/main.js"]

- apps/web/.env.production.example
  - `VITE_API_BASE_URL=http://223.130.160.245` 추가함으로써 `ws://localhost:3000` -> `ws://223.130.160.245/` 로 요청 url 변경
- apps/web/nginx.conf
  - nginx WebSocket 업그레이드 헤더 적용
## 📸 스크린샷 (선택)

<img width="1324" height="815" alt="image" src="https://github.com/user-attachments/assets/2bfcd488-0070-44ba-bfa0-9ca19a753b23" />

## 🧠 의도 및 배경

- 각 팀원이 구현한 내용을 Merge하면서 발생한 버그들을 해결하기 위해 트러블 슈팅을 진행
- 누락된 로직 및 변경되어야 하는 코드를 수정함으로써 코드 동기화, 참가자 동기화를 진행
- 배포 과정에서 소켓 연결에 대해 발생한 문제들을 해결

## 💬 리뷰 요구사항(선택)
- 배포를 완료했기 때문에 CICD까지 구축하면 될 것 같습니다.
- 배포 환경에서 redis 데이터를 날리고 싶으면 종료 및 재빌드 초기화 명령어 참고해서 진행하시면 됩니다. (DOCKER.md 문서 참고)
- `ws://223.130.160.245` url로 요청이 잘 실행됩니다.

eeekeee committed:
fix: 순환 참조 문제 해결 (#19)

- redis.module.ts : redis.service.ts 순환 참조 문제 해결
- redis.service.ts 삭제
- battle-redis.service.ts 분리

cchaeyoung committed:
feat: 방 인원 확인 로직 구현 (#24)

- getRoom: Redis에서 방 조회
- getRoomAvailability: 현재 플레이어 수 및 입장 가능 여부 반환

cchaeyoung committed:
feat: 방 인원 확인 로직 구현 (#24)

- getRoom: Redis에서 방 조회
- getRoomAvailability: 현재 플레이어 수 및 입장 가능 여부 반환

eeekeee committed:
feat: Battle Module 추가 및 벡엔드 path alias 설정 (#19)

- tsconfig.json: paths 추가
- main.ts: import tsconfig-paths/register

eeekeee committed:
feat: Battle Module 추가 및 벡엔드 path alias 설정 (#19)

- tsconfig.json: paths 추가
- main.ts: import tsconfig-paths/register

cchaeyoung committed:
feat: 방 인원 확인 소켓 이벤트 핸들러 구현 (#24)

- CHECK_ROOM_AVAILABILITY 이벤트 핸들러 추가
- client.join()으로 실시간 업데이트 구독
- 현재 인원 정보 응답

cchaeyoung committed:
feat: 방 인원 확인 소켓 이벤트 핸들러 구현 (#24)

- CHECK_ROOM_AVAILABILITY 이벤트 핸들러 추가
- client.join()으로 실시간 업데이트 구독
- 현재 인원 정보 응답

cchaeyoung committed:
feat: 방 저장 메서드 추가 (#25)

- saveRoom: 수정된 Room 객체를 Redis에 저장
- 사용자 입장 시 currentPlayers/Spectators 업데이트용

cchaeyoung committed:
feat: 방 저장 메서드 추가 (#25)

- saveRoom: 수정된 Room 객체를 Redis에 저장
- 사용자 입장 시 currentPlayers/Spectators 업데이트용

cchaeyoung committed:
feat: 방 입장 소켓 이벤트 구현 (#25)

- JOIN_ROOM 이벤트 핸들러 추가
- player/spectator 역할 검증
- RoomUser 생성
- Redis에 방 정보 저장

cchaeyoung committed:
feat: 방 입장 소켓 이벤트 구현 (#25)

- JOIN_ROOM 이벤트 핸들러 추가
- player/spectator 역할 검증
- RoomUser 생성
- Redis에 방 정보 저장

eeekeee committed:
feat: Battle Gateway 생성 및 유저 코드 업데이트 구현 (#19)

- 코드 변경 요청 이벤트: CODE_CHANGE
- 코드 변경 완료 이벤트: CODE_UPDATED

eeekeee committed:
feat: Battle Gateway 생성 및 유저 코드 업데이트 구현 (#19)

- 코드 변경 요청 이벤트: CODE_CHANGE
- 코드 변경 완료 이벤트: CODE_UPDATED

cchaeyoung committed:
feat: 방 입장 시 클라이언트 상태 동기화 및 인원 업데이트 (#26)

- 입장한 유저에게 방 정보 전송
- 기존 유저들에게 인원수 브로드캐스트

cchaeyoung committed:
feat: 방 입장 시 클라이언트 상태 동기화 및 인원 업데이트 (#26)

- 입장한 유저에게 방 정보 전송
- 기존 유저들에게 인원수 브로드캐스트

SubJeeLee committed:
chore: 소켓, zustand 의존성 추가 (#14)
- 소켓 연결/스토어용 패키지 설치 및 lock 업데이트

eeekeee committed:
feat: UpdateUserCodeDTO 수정 (#19)

eeekeee committed:
feat: UpdateUserCodeDTO 수정 (#19)

SubJeeLee committed:
feat: 모달 오픈 시 인원수 요청 및 실시간 반영 (#14,15)
- RoomAvailability DTO 등 타입 추가
- 모달 열 때 바인딩 요청, 카운트 props 계산
- 실시간 roomState를 모달 props로 전달
- 모달 open/close 이벤트 구독 연결/해제

cchaeyoung committed:
refactor: 방 인원 확인 응답 이벤트명에서 CHECK 제거 (#24)

cchaeyoung committed:
refactor: 방 인원 확인 응답 이벤트명에서 CHECK 제거 (#24)

eeekeee committed:
Merge pull request #39 from boostcampwm2025/feat/19-code-event

[FIX] #38

eeekeee committed:
Merge pull request #40 from boostcampwm2025/feat/22-room-entry-role

[FEAT] 방 입장 및 역할 부여 (#22)

dyeon-dev committed:
chore: socket.io, zustand 설치 (#33)

- 클라이언트 소켓 연결을 위한 의존성 설치
- user 정보 저장을 위한 상태관리 설치

eeekeee committed:
feat: JOIN_ROOM DTO 수정 및 ROOM 참가 인원 DTO 구현 (#38)

- CHECK_ROOM_AVAILABILITY 이벤트 응답 데이터 수정
- 소켓 이벤트 상수 변경

eeekeee committed:
feat: JOIN_ROOM DTO 수정 및 ROOM 참가 인원 DTO 구현 (#38)

- CHECK_ROOM_AVAILABILITY 이벤트 응답 데이터 수정
- 소켓 이벤트 상수 변경

dyeon-dev committed:
feat: 코드 에디터 입력 시 소켓 연결 (#33)

- code-change 이벤트로 서버에 전송
- 소켓은 useMemo로 생성
- 언마운트시 useEffect에서 disconnect

SubJeeLee committed:
fix: 실시간 모달 확인 로직 수정 (#14)
- 불필요한 소켓 연결 해제 수정
- response, request 통일

dyeon-dev committed:
feat: 관전자 페이지에서 실시간 코드 동기화 (#33)

- code-updated 이벤트로 서버에서 응답받기
- room-status-sync로 참가자 목록 받기

SubJeeLee committed:
Merge remote-tracking branch 'origin/fix/38-merge' into feat/14-modal-socket

SubJeeLee committed:
feat: 방 상태 DTO 정리 및 모달 실시간 인원 반영 소켓 구독 추가 (#14,15)
- RoomAvailability/JoinRoom DTO 재정의, requestedRole 필드로 백엔드와 정합성 맞춤.
- availability/joined 이벤트 구독 해제 로직 추가, 모달 열림/닫힘에 맞춰 카운트 자동 갱신
- ROOM_USER_LEFT 리스터 추가로 playerCount 업데이트

SubJeeLee committed:
feat: JoinRoom 요청/응답 정리 및 네비게이션 반영 (#16)
- JoinRoom 요청 필드 requestedRole로 통일, 최소 응답 타입으로 단순화
- join 호출 시 ACK/타임아웃 처리 후 배틀 페이지로 이동

eeekeee committed:
feat: 방 생성시 배틀 자동 생성(MVP) (#41)

- createRoom 내부에 createBattle 호출

eeekeee committed:
feat: 유저가 방 입장시 배틀 자동 참가 구현 (#41)

- 참가자일 경우 배틀 참가

eeekeee committed:
feat: leaveBattle 메서드 구현(MVP) (#41)

- 참가자 퇴장 시 배틀에서 제외

eeekeee committed:
Merge branch 'fix/38-merge' into feat/41-room-battle

dyeon-dev committed:
feat: 방 상태 관리를 위한 스토어 추가 (#33)

- ROOM_STATE_SYNC 이벤트를 통해 방 정보를 전역 상태 스토어에 저장
- 현재 참가자, 코드 정보를 스토어에 업데이트

dyeon-dev committed:
feat: 배틀 페이지에 입장하면 전체 상태 동기화하는 이벤트 처리 (#33)

- ROOM_STATE_SYNC 이벤트를 통해 전달받은 데이터를 스토어에 저장

dyeon-dev committed:
feat: 관전자 화면을 스토어 기반으로 전환 (#33)

- CODE_UPDATED가 발생하면 코드 상태 동기화

cchaeyoung committed:
feat: 방 나가기 기능 구현 (#44)

- RoomService.removeUser(): Redis에서 유저 제거 로직 추가
- RoomGateway.handleLeaveRoom(): LEAVE_ROOM 이벤트 핸들러 추가
- ROOM_USER_LEFT 브로드캐스트로 실시간 인원수 업데이트

SubJeeLee committed:
Merge pull request #42 from boostcampwm2025/feat/14-modal-socket

[FEAT] 모달 오픈 시 서버에 현재 인원수 요청 및 실시간 처리 (#14)

eeekeee committed:
Merge pull request #43 from boostcampwm2025/feat/41-room-battle

[FEAT] Room - Battle 자동 연결 (MVP) (#41)

cchaeyoung committed:
Merge pull request #46 from boostcampwm2025/feat/44-room-leave

[FEAT] 방 나가기 기능 구현 (#44)

dyeon-dev committed:
Merge branch 'fix/38-merge' into feat/33-code-socket

dyeon-dev committed:
Merge pull request #45 from boostcampwm2025/feat/33-code-socket

[FEAT] 코드 작성 시 소켓 연결 구현 (#33)

SubJeeLee committed:
hotfix: 서버에서 STATE_SYNC 연결 수정 및 참가자 입장 오류 수정 (#38)

SubJeeLee committed:
hotfix: type 에러 수정 (#38)

eeekeee committed:
fix: Store sokcet 재사용 (#38)

dyeon-dev committed:
fix: 코드 변경 시 관전자 화면에 안나오는 문제 해결 (#38)

- 코드 이벤트는 CODE_CHANGE와 CODE_UPDATED 둘 다 수신해 동일 핸들러로 upsertCode를 갱신
- ROOM_STATE_SYNC 이벤트도 구독해서 upsertPlayer, setMe로 스토어에 참가자/자신을 넣어 players 목록 생성

dyeon-dev committed:
chore: tsconfig-paths 의존성 devDependencies에서 dependencies으로 변경 (#1)

- 빌드 결과물(main.js)이 tsconfig-paths/register를 require하고 있음
- 하지만 production 이미지에는 해당 패키지가 없음
- tsconfig-paths를 prod에서도 설치하도록 devDependencies → dependencies로 이동

dyeon-dev committed:
fix: room.gateway 이전 코드로 롤백 (#38)

- 버그 수정을 위해 임시로 변경해둔 코드 롤백

dyeon-dev committed:
chore: 업데이트된 pnpm-lock.yaml 추가 (#1)

dyeon-dev committed:
chore: api dockerfile 빌드 산출물 경로 변경 (#1)

- 배포환경 빌드 시 산출물 경로 안맞는 문제 해결

dyeon-dev committed:
chore: nginx WebSocket 업그레이드 헤더 적용 (#38)

- HTTP 연결은 되는데 socket.io 클라이언트가 무한 reconnect되는 문제 해결

dyeon-dev committed:
chore: 배포 환경 socket 연결 ip 주소 할당 (#38)

- localhost:3000으로 socket 연결되는 문제 해결

dyeon-dev committed:
docs: redis 초기화 명령어 정의 (#38)

dyeon-dev committed:
fix: stores 폴더로 roomStore 파일 이동 (#38)

eeekeee reviewed (approved):
배포가 되어서 다행이군요! 고생하셨습니다!

참가자 목록과 나가기 연동은 추후에 학인해보겠습니다

eeekeee committed:
feat: 방 나가기 기능 UI와 연동 (#48)

- Store에 leaveRoom 함수 구현
- BattleHeader의 나가기 버튼 onClick 연결

dyeon-dev committed:
fix: test 폴더로 spec 파일 이동 (#38)

dyeon-dev committed:
fix: 기존 ROOM_STATE_SYNC를 역할 부여 이벤트로 변경 및 전체 참가자 목록 이벤트 추가 (#38)

- ROOM_STATE_ROLE: joinRoom에서 방 상태 동기화가 아닌 역할에 따라 응답값을 주는 이벤트이기 때문에 이벤트 용도를 명확하게 하기 위해 네임 변경
- ROOM_PLAYERS: 방 참가자 목록을 위한 이벤트 추가

dyeon-dev committed:
feat: 방 참가자 목록 DTO 추가 (#38)

- room.ts: RoomPlayerPayload
- user.ts: 방에 들어간 사용자 roomId 데이터 추가

dyeon-dev committed:
feat: 방 참여자 목록 브로드 캐스트 추가 (#38)

- ROOM_PLAYERS 이벤트 브로드 캐스트

dyeon-dev committed:
feat: 관전자 페이지에서 스토어에 업데이트된 참가자 목록 동기화 (#38)

- joinRoom 단계에서 ROOM_PLAYERS를 스토어에 반영하도록 추가

dyeon-dev committed:
fix: 참가자 목록 동기화에 필요한 로직 수정 (#38)

- joinRoom을 실행하는 곳에서 참가자를 가져오기 때문에 CodeSpectator, CodeEditor 부분에서는 필요 없음

cchaeyoung reviewed (approved):
소켓 동기화 문제가 잘 해결되었네요!

수고하셨습니다~!!

SubJeeLee reviewed (approved):
인터미션 기간동안 수고 많으셨습니다!

참고해야할 내용들도 PR에 자세히 적어주셔서 좋았습니다!
